Add /display documentation

m-e-l-u-h-a-n opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
It adds documentation for `/display` settings in `Zulip api documentation`. Links for `/display` documentation has been added to `/register` documentation as asked in issue #16048 

fixes: #16048 

**Testing plan:**  `./tools/check-openapi`,  `./tools/lint`, `./tools/test-api`



<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

m-e-l-u-h-a-n commented on zerver/openapi/zulip.yaml:
```diff
@@ -7955,6 +7979,203 @@ paths:
+          schema:
+            type: boolean
+          example: true
+        - name: dense_mode
+          in: query
+          description: |
+            Set display in dense mode.
+            It is an experimental setting only available in development environments.
+          schema:
+            type: boolean
+          example: true
```
I could not find this setting as well as `high_contrast_mode` setting available on czo, but these were available on development server. After going through various documentation and code related to these settings I found that `dense_mode` is an experimental settings only available in development environment. So I have added the same description for `dense_mode`. But as I could not find any such description for `high_contrast_mode` anywhere I have added similar description for this also. So any suggestions for change over documentation these two settings are highly awaited.

m-e-l-u-h-a-n commented:
@timabbott can you please give it a review.

timabbott commented:
As a natural follow-up, I think we should look at https://github.com/zulip/zulip/issues/15307 (and https://chat.zulip.org/#narrow/stream/3-backend/topic/update-subscription-properties/near/901757).  

@akashaviator do you have any work on this from last summer that we should build off of?

timabbott commented on templates/zerver/api/update-display-settings.md:
```diff
@@ -0,0 +1,35 @@
+The server will return the settings that have been changed after the request,
+with their new value. Please note that this doesn't necessarily mean that it
+will return all the settings passed as parameters in the request, but only
+those ones that were different than the already existing setting.
```
Structurally, this sort of paragraph isn't how we want our docs to work but see my comment about planned changes to this endpoint's structure that would rewrite how the return values work; given that plan, I think it's probably better to merge this roughly as-is. (I'll do a bit of copyediting).

m-e-l-u-h-a-n replied:
Do we want that format(mentioned on CZO) to be documented here now or,
Should I wait untill the plan you mentioned in [https://chat.zulip.org/#narrow/stream/3-backend/topic/update-subscription-properties/near/901757](https://chat.zulip.org/#narrow/stream/3-backend/topic/update-subscription-properties/near/901757) gets implemented?

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -8016,6 +8040,205 @@ paths:
+            application/json:
+              schema:
+                type: string
+              example: "Asia/Kolkata"
```
For all of these settings, can you add a link in the description to the appropriate part of the `/help/` documentation covering the setting in question?  This sort of cross-linking is really important to making our API documentation useful.

m-e-l-u-h-a-n replied:
I made this changes although I am not 100% sure if the sentence I have added is correct. This is the [commit](https://github.com/zulip/zulip/pull/17033/commits/134aa4eaeafc11dfe4eb9a7207c625e030775cfd) that adds link for this /display api documentation in corresponding help docs. Please have a look.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -6732,6 +6754,8 @@ paths:
                           Only relevant to clients with configuration UI for choosing an emojiset;
                           the currently selected emojiset is available in the `emojiset` key.
+                          See [update-display-settings](/api/update-display-settings)
+                          for details on display settings.
```
Please add a blank line before your new paragraphs, for consistency.

timabbott replied:
Also, we name endpoints in these docs via e.g. `POST /...`, not the URL chunks, if we reference them.  I think cleanest though would be to just see "See the [API documentation for changing display settings](/api/update-display-settings)".

Also, you should reverse the order of the two commits, because the first commit has broken links without the second merged.

m-e-l-u-h-a-n replied:
Made these changes, I have put the links in following pattern:
`[PATCH /update-display-settings](/api/update-display-settings)`

m-e-l-u-h-a-n replied:
Corrected the commit order too.

timabbott commented:
This looks pretty good; I posted a batch of mostly minor comments.  Thanks for all your work on this @m-e-l-u-h-a-n!

akashaviator commented:
@timabbott #15317 We started some prototyping here.

m-e-l-u-h-a-n commented:
@timabbott I have addressed all of your comments on this pr, except for [this comment](https://github.com/zulip/zulip/pull/17033#discussion_r581486276) of yours for which I was not very much sure of. Here is what I changed:
* Add links for `/display` documentation in corresponding `/help` pages.
* Changed the pattern in which links are mentioned, it is now `[PATCH /update-display-settings](/api/update-display-settings)`.
* correction of commit order.

I am not 100% sure of the sentences I added in `/help` documentation. I added them now just to have some work over which we can discuss. Please have a look.

m-e-l-u-h-a-n commented:
I found that there were no links to api documentation for notifications settings in `/help/` documentation. Should I add them to?

abhijeetbodas2001 commented on templates/zerver/api/update-display-settings.md:
```diff
@@ -0,0 +1,35 @@
+The server will return the settings that have been changed after the request,
+with their new value. Please note that this doesn't necessarily mean that it
+will return all the settings passed as parameters in the request, but only
+those ones that were different than the already existing setting.
```
Should this be "... different *from*"? Reference: https://english.stackexchange.com/a/845. In the codebase too, there are ~28 of "different from" and just [one](https://github.com/zulip/zulip/blob/master/templates/zerver/api/update-notification-settings.md) of "different than".

timabbott replied:
yes, agreed.

m-e-l-u-h-a-n replied:
Made these changes.

timabbott commented:
@m-e-l-u-h-a-n let me know when you have this rebased.  

And thanks @akashaviator for the review :)

m-e-l-u-h-a-n commented:
@timabbott I have rebased this pr.
Also documented new `default_view` setting.

timabbott commented:
I'm dropping the last commit -- we don't to link from /help to /api, and that's probably unnecessary in general -- the links we want are in the other direction (so one can understand how fields in the API documentation relate to user-facing behavior).

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -6756,37 +6756,55 @@ paths:
                           The color scheme selected by the user.
+
+                          See [PATCH /update-display-settings](/api/update-display-settings)
+                          for api details on display settings.
```
We have a standard format for linking to other API endpoints, and it's this:

```
diff --git a/zerver/openapi/zulip.yaml b/zerver/openapi/zulip.yaml
index a7d6fad357..23ca5b6cb4 100644
--- a/zerver/openapi/zulip.yaml
+++ b/zerver/openapi/zulip.yaml
@@ -7556,7 +7556,7 @@ paths:
 
                           The color scheme selected by the user.
 
-                          See [PATCH /update-display-settings](/api/update-display-settings)
+                          See [PATCH /settings/display](/api/update-display-settings)
                           for api details on display settings.
                       default_language:
                         type: string
```

I kinda want to extend the markdown process so that we can write something like `[update-display-settings]` and have the right thing happen (we have the data to do that) for these links.

Ideally such a system would also let us do `[feature-level-49]` and auto-generated the text with a link to the API changelog as well.  

@orientor @MSurfer20 what do you think of this idea?  I'm not sure what syntax would make sense, but it does feel like it could simplify the experience for adding links consistently.

m-e-l-u-h-a-n replied:
I'll update it if you are not looking at it right now.

timabbott replied:
For this PR, I decided this was close enough to finishing that it'd be better for me to just do the quick find-replace.

But I'd appreciate feedback on the markdown feature; we write a lot of these links, especially in the API changelog, and they're kinda fiddly, so it'd be nice if we could remove the fiddliness and need for contributors to be consistent with how we do the labels (etc.).

Can one of you start a thread on #documentation about this idea?

abhijeetbodas2001 replied:
(For reference, pasting here the link to the chat topic opened by Purushottam 🙂 )
https://chat.zulip.org/#narrow/stream/19-documentation/topic/Link.20.2Fhelp.20page.20using.20template-name

timabbott commented:
I went ahead and fixed this so I could merge it without another round trip.  Changes I made include:
* Rewriting the descriptions of what parameters do -- this I expected to do in any case for this sort of PR, since I know the exact status of these better than contributors.  
* Adding links from the documentation to /help/ as discussed while doing so.
* Fixing the string encoding for the "google" and one other example -- though this bug in your OpenAPI content was masked by c6f27440c65b78c9f72ef1b11d045b6e3f7ca34a.
* Adjusted the link format as mentioned above: https://github.com/zulip/zulip/pull/17033#discussion_r608889172
* Adding the `Fixes: ` line to the commit message to close the original issue -- though I think that's just because I dropped the commit where you'd had this.

Thanks @m-e-l-u-h-a-n!
